crates/bevy_math/src/primitives/dim3.rs

```diff
@@ -629,6 +632,163 @@ impl Torus {
+                Vec3::new(0.5, -0.5, 0.0),
+            ],
+        }
+    }
+}
+
+impl Triangle3d {
+    /// Create a new [`Triangle3d`] from points `a`, `b`, and `c`.
+    #[inline(always)]
+    pub fn new(a: Vec3, b: Vec3, c: Vec3) -> Self {
+        Self {
+            vertices: [a, b, c],
+        }
+    }
+
+    /// Get the area of the triangle.
+    #[inline(always)]
+    pub fn area(&self) -> f32 {
+        let [a, b, c] = self.vertices;
+        let ab = b - a;
+        let ac = c - a;
+        ab.cross(ac).length() / 2.0
+    }
+
+    /// Get the perimeter of the triangle.
+    #[inline(always)]
+    pub fn perimeter(&self) -> f32 {
+        let [a, b, c] = self.vertices;
+        a.distance(b) + b.distance(c) + c.distance(a)
+    }
+
+    /// Get the normal of the triangle in the direction of the right-hand rule, assuming
+    /// the vertices are ordered in a counter-clockwise direction.
+    ///
+    /// The normal is computed as the cross product of the vectors `ab` and `ac`.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err(InvalidDirectionError)`](InvalidDirectionError) if the length
+    /// of the given vector is zero (or very close to zero), infinite, or `NaN`.
+    #[inline(always)]
+    pub fn normal(&self) -> Result<Dir3, InvalidDirectionError> {
+        let [a, b, c] = self.vertices;
+        let ab = b - a;
+        let ac = c - a;
+        Dir3::new(ab.cross(ac))
+    }
+
+    /// Checks if the triangle is degenerate, meaning it has zero area.
+    ///
+    /// A triangle is degenerate if the cross product of the vectors `ab` and `ac` has a length less than `f32::EPSILON`.
+    /// This indicates that the three vertices are collinear or nearly collinear.
+    #[inline(always)]
+    pub fn is_degenerate(&self) -> bool {
+        let [a, b, c] = self.vertices;
+        let ab = b - a;
+        let ac = c - a;
+        ab.cross(ac).length() < f32::EPSILON
+    }
+
+    /// Reverse the triangle by swapping the first and last vertices.
+    #[inline(always)]
+    pub fn reverse(&mut self) {
+        self.vertices.swap(0, 2);
+    }
+
+    /// Get the centroid of the triangle.
+    ///
+    /// This function finds the geometric center of the triangle by averaging the vertices:
+    /// `centroid = (a + b + c) / 3`.
+    #[doc(alias("center", "barycenter", "baricenter"))]
+    #[inline(always)]
+    pub fn centroid(&self) -> Vec3 {
+        (self.vertices[0] + self.vertices[1] + self.vertices[2]) / 3.0
+    }
+
+    /// Get the circumcenter of the triangle.
+    #[inline(always)]
+    pub fn circumcenter(&self) -> Vec3 {
+        let [a, b, c] = self.vertices;
+        let ab = b - a;
+        let ac = c - a;
+        // Reference: https://gamedev.stackexchange.com/questions/60630/how-do-i-find-the-circumcenter-of-a-triangle-in-3d
+        let n = ab.cross(ac);
+        a + ((ac.length_squared() * n.cross(ab) + ab.length_squared() * ac.cross(ab).cross(ac))
+            / (2.0 * n.length_squared()))
```

Comment:
Division by 0 for degenerate triangles happens here, not sure how to resolve it.

Reply:
Discussed elsewhere, should default to the midpoint of the line if verts are colinear, or the vertex if all verts are identical. Right?

Reply:
If all verts are identical I agree you should produce a circle of radius 0 and a center in the vertices.
If two vertices are identical, it should be the midpoint of the nonzero line segment.
If the vertices are distinct but collinear, I'd say there should be no circumcircle (or a circle at infinity??)

Reply:
It appears that there are three cases for collinear vertices:

1. `A = B = C`
2. `A = B`, and `B != C`
3. `A != B != C`

Here's the solution I will implement for each one:

1. `circumcenter` returns `A`
2. `circumcenter` returns the midpoint of the line segment
3. `circumcenter` returns the midpoint of the line segment

Do you agree?

Reply:
Which line segment do you choose in option 3?

Reply:
The largest segment? In general you can find the largest edge segment and return the midpoint of that, for all three cases. I don't see any obvious correctness issues with that.

Reply:
Yes, the largest segment.

> In general you can find the largest edge segment and return the midpoint of that

Good point.

Reply:
I am not sure if I like how I handled it: [commit](https://github.com/bevyengine/bevy/pull/12508/commits/2327c2a676f505b664f15d07bdd5957604b5f80b).

It's now making some calculations twice.